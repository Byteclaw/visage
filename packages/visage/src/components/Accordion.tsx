import { ExtractVisageComponentProps } from '@byteclaw/visage-core';
import React, {
  Children,
  cloneElement,
  ReactElement,
  ReactNode,
  useState,
  MouseEventHandler,
  MouseEvent,
  KeyboardEventHandler,
  KeyboardEvent,
} from 'react';
import { createComponent } from '../core';
import {
  getNextIndexFromCycle,
  createControlActiveShadow,
  createControlFocusShadow,
} from './shared';
import { booleanVariant } from '../variants';
import { useUniqueId } from '../hooks';

const AccordionTrigger = createComponent('div', {
  displayName: 'AccordionTrigger',
  styles: {
    borderColor: 'transparent',
    borderWidth: 0,
    backgroundColor: 'transparent',
    borderBottomColor: 'neutral',
    borderBottomWidth: 1,
    borderStyle: 'solid',
    cursor: 'pointer',
    display: 'flex',
    fontSize: 1,
    fontFamily: 'heading',
    fontWeight: 'bolder',
    lineHeight: 1,
    outline: 'none',
    m: 0,
    p: 0,
    py: 1,
    width: '100%',
    '&:focus': {
      boxShadow: createControlFocusShadow(),
    },
    '&:focus:active': {
      boxShadow: createControlActiveShadow(),
    },
  },
  defaultProps: { role: 'button' },
});

type AccordionTriggerProps = ExtractVisageComponentProps<
  typeof AccordionTrigger
>;

const AccordionRegion = createComponent('div', {
  displayName: 'AccordionRegion',
  styles: props => ({
    height: props.open ? 'auto' : 0,
    opacity: props.open ? 1 : 0,
    m: 0,
    p: 0,
    py: props.open ? 1 : 0,
    visibility: props.open ? 'visible' : 'hidden',
    transition:
      'height cubic-bezier(0.2, 0, 0.38, 0.9) 110ms, padding cubic-bezier(0.2, 0, 0.38, 0.9) 110ms',
  }),
  defaultProps: { role: 'region' },
  variants: [booleanVariant('open', true)],
});

type AccordionRegionProps = ExtractVisageComponentProps<typeof AccordionRegion>;

export interface AccordionItemProps {
  children: ReactNode | (() => ReactNode);
  open?: boolean;
  onTriggerClick?: MouseEventHandler<HTMLDivElement>;
  onTriggerKeyDown?: KeyboardEventHandler<HTMLDivElement>;
  title: ReactNode;
  regionId?: string;
  regionProps?: AccordionRegionProps;
  triggerId?: string;
  triggerProps?: AccordionTriggerProps;
}

export function AccordionItem({
  children,
  onTriggerClick,
  onTriggerKeyDown,
  open,
  // passed from Accordion
  regionId,
  regionProps,
  title,
  // passed from Accordion
  triggerId,
  triggerProps,
}: AccordionItemProps) {
  return (
    <>
      <AccordionTrigger
        {...regionProps}
        aria-controls={regionId}
        aria-expanded={open}
        id={triggerId}
        onClick={onTriggerClick}
        onKeyDown={onTriggerKeyDown}
        tabIndex={0}
      >
        {title}
      </AccordionTrigger>
      <AccordionRegion
        {...triggerProps}
        aria-labelledby={triggerId}
        id={regionId}
        open={open}
      >
        {typeof children === 'function' ? children() : children}
      </AccordionRegion>
    </>
  );
}

type AccordionItemElement = ReactElement<AccordionItemProps>;

interface AccordionProps {
  id?: string;
  children: AccordionItemElement | AccordionItemElement[];
}

export function Accordion({ children, id }: AccordionProps) {
  const accordionId = useUniqueId(id, 'accordion');

  const items: ReactElement<AccordionItemProps>[] = Children.toArray(
    children,
  ) as ReactElement<AccordionItemProps>[];

  const [openItem, setOpenItem] = useState(() => {
    const idx = items.findIndex(item => item.props.open);

    return idx === -1 ? 0 : 1;
  });

  // if out of bounds
  if (openItem >= items.length || openItem < 0) {
    setOpenItem(0);
  }

  return (
    <>
      {items.map((item, i) =>
        cloneElement(item, {
          regionId: `${accordionId}-${i}-region`,
          triggerId: `${accordionId}-${i}-trigger`,
          open: openItem === i,
          onTriggerClick: (e: MouseEvent<HTMLDivElement>) => {
            setOpenItem(i);

            if (item.props.onTriggerClick) {
              item.props.onTriggerClick(e);
            }
          },
          onTriggerKeyDown: (e: KeyboardEvent<HTMLDivElement>) => {
            if (e.key === 'Enter' || e.key === ' ') {
              e.preventDefault();

              setOpenItem(i);
            } else if (e.key === 'ArrowUp' || e.key === 'ArrowDown') {
              e.preventDefault();

              const itemToFocus = document.getElementById(
                `${accordionId}-${getNextIndexFromCycle(
                  i,
                  e.key === 'ArrowUp' ? -1 : 1,
                  items.length - 1,
                )}-trigger`,
              );

              if (itemToFocus) {
                itemToFocus.focus();
              }
            } else if (e.key === 'Home' || e.key === 'End') {
              e.preventDefault();

              const itemToFocus = document.getElementById(
                `${accordionId}-${
                  e.key === 'Home' ? 0 : items.length - 1
                }-trigger`,
              );

              if (itemToFocus) {
                itemToFocus.focus();
              }
            }

            if (item.props.onTriggerKeyDown) {
              item.props.onTriggerKeyDown(e);
            }
          },
        }),
      )}
    </>
  );
}
