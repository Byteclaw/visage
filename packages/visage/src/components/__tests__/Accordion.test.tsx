import React from 'react';
import { fireEvent } from '@testing-library/react';
import { render, createTestTheme } from './render';
import { Accordion, AccordionItem } from '../Accordion';

describe('Accordion', () => {
  it('renders correctly', () => {
    const { asFragment } = render(
      <Accordion>
        <AccordionItem title="A">A</AccordionItem>
      </Accordion>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('reacts to keyboard events', () => {
    const { container } = render(
      <Accordion>
        <AccordionItem title="A">A</AccordionItem>
        <AccordionItem title="B">B</AccordionItem>
        <AccordionItem title="C">C</AccordionItem>
      </Accordion>,
    );
    const A = container.querySelector(
      '#accordion-1-0-trigger',
    ) as HTMLDivElement;
    const B = container.querySelector(
      '#accordion-1-1-trigger',
    ) as HTMLDivElement;
    const C = container.querySelector(
      '#accordion-1-2-trigger',
    ) as HTMLDivElement;

    // A should be expanded by default
    expect(A).toHaveAttribute('aria-expanded', 'true');

    A.focus();

    expect(A).toHaveFocus();

    // focus B
    B.focus();

    expect(B).toHaveFocus();

    // expand B by Enter keydown
    fireEvent.keyDown(B, {
      key: 'Enter',
    });

    expect(A).toHaveAttribute('aria-expanded', 'false');
    expect(B).toHaveAttribute('aria-expanded', 'true');
    expect(C).toHaveAttribute('aria-expanded', 'false');

    // press space, does nothing
    fireEvent.keyDown(B, {
      key: ' ',
    });

    expect(B).toHaveAttribute('aria-expanded', 'true');

    // focus next using arrow down
    fireEvent.keyDown(B, {
      key: 'ArrowDown',
    });

    expect(C).toHaveFocus();

    // now expand using space
    // focus next using arrow down
    fireEvent.keyDown(C, {
      key: ' ',
    });

    expect(C).toHaveAttribute('aria-expanded', 'true');

    // arrow down goes back to A
    fireEvent.keyDown(C, {
      key: 'ArrowDown',
    });

    expect(A).toHaveFocus();

    // arrow up returns back to C
    fireEvent.keyDown(A, {
      key: 'ArrowUp',
    });

    expect(C).toHaveFocus();

    // Home returns back to A
    fireEvent.keyDown(C, {
      key: 'Home',
    });

    expect(A).toHaveFocus();

    // End returns back to C
    fireEvent.keyDown(A, {
      key: 'End',
    });

    expect(C).toHaveFocus();
  });

  it('reacts to mouse events', () => {
    const { container } = render(
      <Accordion>
        {['A', 'B', 'C'].map(title => (
          <AccordionItem key={title} title={title}>
            {title}
          </AccordionItem>
        ))}
      </Accordion>,
    );
    const A = container.querySelector('#accordion-1-0-trigger');
    const B = container.querySelector('#accordion-1-1-trigger');
    const C = container.querySelector('#accordion-1-2-trigger');

    // A should be expanded by default
    expect(A).toHaveAttribute('aria-expanded', 'true');

    fireEvent.click(B);

    expect(A).toHaveAttribute('aria-expanded', 'false');
    expect(B).toHaveAttribute('aria-expanded', 'true');
    expect(C).toHaveAttribute('aria-expanded', 'false');
  });

  it('can be styles using AccordionRegion face', () => {
    const { asFragment } = render(
      <Accordion>
        <AccordionItem title="A">A</AccordionItem>
      </Accordion>,
      {
        ds: {
          theme: createTestTheme({
            AccordionRegion: {
              color: 'blue',
            },
          }),
        },
      },
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('can be styles using AccordionTrigger face', () => {
    const { asFragment } = render(
      <Accordion>
        <AccordionItem title="A">A</AccordionItem>
      </Accordion>,
      {
        ds: {
          theme: createTestTheme({
            AccordionTrigger: {
              color: 'blue',
            },
          }),
        },
      },
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
