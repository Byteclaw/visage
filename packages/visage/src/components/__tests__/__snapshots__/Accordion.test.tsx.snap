// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion can be styles using AccordionRegion face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-0 {
  border-color: rgba(0,0,0,0);
  border-width: 0;
  background-color: rgba(0,0,0,0);
  border-bottom-color: rgb(238,238,238);
  border-bottom-width: 1px;
  border-style: solid;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 26px;
  font-family: heading-font;
  font-weight: bolder;
  line-height: 48px;
  outline: none;
  margin: 0px;
  padding: 0px;
  padding-top: 8px;
  padding-bottom: 8px;
  width: 100%;
}

.emotion-0:focus {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255);
}

.emotion-0:focus:active {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255) ,inset 0 0 200px rgba(68,68,68,0.2),inset 0 0 2px rgba(0,0,0,0.2);
}

<div
    aria-controls="accordion-1-0-region"
    aria-expanded="true"
    class="emotion-0"
    id="accordion-1-0-trigger"
    role="button"
    tabindex="0"
  >
    A
  </div>
  .emotion-0 {
  height: auto;
  opacity: 1;
  margin: 0px;
  padding: 0px;
  padding-top: 8px;
  padding-bottom: 8px;
  visibility: visible;
  -webkit-transition: height cubic-bezier(0.2,0,0.38,0.9) 110ms,padding cubic-bezier(0.2,0,0.38,0.9) 110ms;
  transition: height cubic-bezier(0.2,0,0.38,0.9) 110ms,padding cubic-bezier(0.2,0,0.38,0.9) 110ms;
  color: rgb(0,0,255);
}

<div
    aria-labelledby="accordion-1-0-trigger"
    class="emotion-0"
    data-open="true"
    id="accordion-1-0-region"
    role="region"
  >
    A
  </div>
</DocumentFragment>
`;

exports[`Accordion can be styles using AccordionTrigger face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-0 {
  border-color: rgba(0,0,0,0);
  border-width: 0;
  background-color: rgba(0,0,0,0);
  border-bottom-color: rgb(238,238,238);
  border-bottom-width: 1px;
  border-style: solid;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 26px;
  font-family: heading-font;
  font-weight: bolder;
  line-height: 48px;
  outline: none;
  margin: 0px;
  padding: 0px;
  padding-top: 8px;
  padding-bottom: 8px;
  width: 100%;
  color: rgb(0,0,255);
}

.emotion-0:focus {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255);
}

.emotion-0:focus:active {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255) ,inset 0 0 200px rgba(68,68,68,0.2),inset 0 0 2px rgba(0,0,0,0.2);
}

<div
    aria-controls="accordion-1-0-region"
    aria-expanded="true"
    class="emotion-0"
    id="accordion-1-0-trigger"
    role="button"
    tabindex="0"
  >
    A
  </div>
  .emotion-0 {
  height: auto;
  opacity: 1;
  margin: 0px;
  padding: 0px;
  padding-top: 8px;
  padding-bottom: 8px;
  visibility: visible;
  -webkit-transition: height cubic-bezier(0.2,0,0.38,0.9) 110ms,padding cubic-bezier(0.2,0,0.38,0.9) 110ms;
  transition: height cubic-bezier(0.2,0,0.38,0.9) 110ms,padding cubic-bezier(0.2,0,0.38,0.9) 110ms;
}

<div
    aria-labelledby="accordion-1-0-trigger"
    class="emotion-0"
    data-open="true"
    id="accordion-1-0-region"
    role="region"
  >
    A
  </div>
</DocumentFragment>
`;

exports[`Accordion renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-0 {
  border-color: rgba(0,0,0,0);
  border-width: 0;
  background-color: rgba(0,0,0,0);
  border-bottom-color: rgb(238,238,238);
  border-bottom-width: 1px;
  border-style: solid;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 26px;
  font-family: heading-font;
  font-weight: bolder;
  line-height: 48px;
  outline: none;
  margin: 0px;
  padding: 0px;
  padding-top: 8px;
  padding-bottom: 8px;
  width: 100%;
}

.emotion-0:focus {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255);
}

.emotion-0:focus:active {
  box-shadow: 0 0 0 4px rgba(255,255,255,0.4),0 0 0 4px rgb(0,0,255) ,inset 0 0 200px rgba(68,68,68,0.2),inset 0 0 2px rgba(0,0,0,0.2);
}

<div
    aria-controls="accordion-1-0-region"
    aria-expanded="true"
    class="emotion-0"
    id="accordion-1-0-trigger"
    role="button"
    tabindex="0"
  >
    A
  </div>
  .emotion-0 {
  height: auto;
  opacity: 1;
  margin: 0px;
  padding: 0px;
  padding-top: 8px;
  padding-bottom: 8px;
  visibility: visible;
  -webkit-transition: height cubic-bezier(0.2,0,0.38,0.9) 110ms,padding cubic-bezier(0.2,0,0.38,0.9) 110ms;
  transition: height cubic-bezier(0.2,0,0.38,0.9) 110ms,padding cubic-bezier(0.2,0,0.38,0.9) 110ms;
}

<div
    aria-labelledby="accordion-1-0-trigger"
    class="emotion-0"
    data-open="true"
    id="accordion-1-0-region"
    role="region"
  >
    A
  </div>
</DocumentFragment>
`;
